xmtp_mls/src/identity.rs

```diff
@@ -639,3 +671,193 @@ pub fn parse_credential(credential_bytes: &[u8]) -> Result<InboxId, IdentityErro
     let cred = MlsCredential::decode(credential_bytes)?;
     Ok(cred.inbox_id)
 }
+
+pub fn build_post_quantum_public_key_extension(
+    public_key: &[u8],
+) -> Result<Extension, IdentityError> {
+    let ext =
+        WrapperEncryptionExtension::new(WrapperAlgorithm::XWingMLKEM768Draft6, public_key.to_vec());
+
+    Ok(ext.try_into()?)
+}
+#[derive(Debug, Error)]
+pub enum GeneratePostQuantumKeyError {
+    #[error(transparent)]
+    Crypto(#[from] openmls_traits::types::CryptoError),
+    #[error(transparent)]
+    Rand(#[from] openmls_libcrux_crypto::RandError),
+}
+
+pub(crate) fn generate_post_quantum_key() -> Result<HpkeKeyPair, GeneratePostQuantumKeyError> {
+    let provider = LibcruxProvider::default();
+    let rand = provider.rand();
+
+    let ikm: SecretVLBytes = rand
+        .random_vec(POST_QUANTUM_CIPHERSUITE.hash_length())?
+        .into();
+
+    Ok(provider
+        .crypto()
+        .derive_hpke_keypair(POST_QUANTUM_CIPHERSUITE.hpke_config(), ikm.as_slice())?)
+}
+
+// Store the hash reference, keyed with both the public init key and the post quantum init key.
+// This is needed to get to the private key when decrypting welcome messages.
+fn store_key_package_references(
+    provider: impl MlsProviderExt,
+    kp: &KeyPackage,
+    // The post quantum init key for the key package used for Post Quantum Welcome Wrapper encryption
+    post_quantum_keypair: &HpkeKeyPair,
+) -> Result<(), IdentityError> {
+    let public_init_key = kp.hpke_init_key().tls_serialize_detached()?;
+
+    let hash_ref = serialize_key_package_hash_ref(kp, &provider)?;
+    let post_quantum_public_key = pq_key_package_references_key(&post_quantum_keypair.public)?;
+
+    // We need to store this in a bincode friendly format so that `read` will work later.
+    // TODO:(nm) review whether this breaks the Zeroize guarantees
```

Comment:
What Zeroize guarantees are we worried about here?

Reply:
That the private key gets zeroed after use. Not the most pressing concern, but something we should at least keep track of so that we can do a pass later and zeroize everything we can across our code. OpenMLS is already very diligent for their structs.